src/Sylius/Bundle/ApiBundle/Validator/Constraints/PromotionCouponEligibilityValidator.php

```diff
@@ -43,29 +47,44 @@ public function validate(mixed $value, Constraint $constraint): void
             return;
         }
 
+        /** @var PromotionCouponInterface|null $promotionCoupon */
         $promotionCoupon = $this->promotionCouponRepository->findOneBy(['code' => $value->couponCode]);
 
-        if (!$promotionCoupon instanceof PromotionCouponInterface) {
-            $this->context
-                ->buildViolation($constraint->message)
-                ->atPath('couponCode')
-                ->addViolation()
-            ;
+        if ($promotionCoupon === null) {
+            $this->addViolation($constraint->invalid, 'COUPON_INVALID');
 
             return;
         }
 
-        /** @var OrderInterface $cart */
+        $expirationDate = $promotionCoupon->getExpiresAt();
+
+        if ($expirationDate !== null && $expirationDate < new \DateTime()) {
+            $this->addViolation($constraint->expired, 'COUPON_EXPIRED');
+
+            return;
+        }
+
+        /** @var OrderInterface|null $cart */
         $cart = $this->orderRepository->findCartByTokenValue($value->orderTokenValue);
 
+        if ($cart === null) {
+            throw new \InvalidArgumentException('Cart with given token supposed to exist at this point.');
+        }
+
         $cart->setPromotionCoupon($promotionCoupon);
 
         if (!$this->appliedCouponEligibilityChecker->isEligible($promotionCoupon, $cart)) {
-            $this->context
-                ->buildViolation($constraint->message)
-                ->atPath('couponCode')
-                ->addViolation()
-            ;
+            $this->addViolation($constraint->ineligible, 'PROMOTION_INELIGIBLE');
```

Comment:
Since this is coupon eligibility checker, code should be `COUPON_INELIGIBLE`.

But I would suggest to use fill `PROMOTION_COUPON_INELIGIBLE` name.

```suggestion
            $this->addViolation($constraint->ineligibleMessage, PromotionCouponEligibility::PROMOTION_COUPON_INELIGIBLE_ERROR);
```
Following `symfony/validation` notation, error code const should be `ACME_ERROR`
![image](https://github.com/user-attachments/assets/aa7054dc-3646-4656-b1b0-7c9c11b3cf5c)
![image](https://github.com/user-attachments/assets/5ee5845c-f37f-464b-aef3-d6833a19d82a)